### DISCUSSION TITLE
Curly brackets instead of :

### DISCUSSION THREAD

**Discussion by Omicronlawful** [modular/modular#258] [Ideas]
The option to use curly brackets instead of indents and :. The option to use both would make it easier to code and format the code. Some more python code gets extremly hard to read when there are too many indents and auto formatting isn’t working that well when you look at cargo fmt. It’s pretty useful to just write the code and format it later. It happens sometimes in rust that you mess up the indents, which will be fixed by the formatter later. A wrong indent will break python code and that happens fast.

**Comment by daydun** [comment]
This would be neat, but I also understand if it isn't added.

There are existing languages that allow both indentation based blocks, as well as whitespace agnostic blocks surrounded in curly braces, so it is possible to pull off. Though one would have to consider if it clashes with any other Python syntax.

**Reply by Omicronlawful** [reply]
It could be added as a feature through another filetype or a comment at the top ‘#use_bracket_syntax’. The parser would need to be modified but in comparison to the rest of the project it would be easy

**Reply by daydun** [reply]
That does not sound like a very elegant solution.

**Comment by Mogball** [comment]
There are practical reasons why brackets will not work and why significant whitespace is crucial to the parser: lazy body parsing. Mojo's parser can trivially skip over the body of structs, functions, etc. because it can use the expected indentation to find the end of the indentation block.

**Reply by daydun** [reply]
Why couldn't you just parse the body into AST form and then skip type checking and everything until it's needed? Does parsing depend on type information?

**Reply by Omicronlawful** [reply]
I think this is only relevant for jit. When the code is compiled before it does only affect the performance of the compiler which does not really matter in normal sized projects. For jit it could just convert the brackets to the python syntax.  Formatting the file and replacing the curly braces takes only a few ms when compiling.

**Reply by Mogball** [reply]
> Does parsing depend on type information? 

yes

**Reply by Omicronlawful** [reply]
> > Does parsing depend on type information?
> 
> yes

What about the attempt to implement that feature as a formatter that formats when compiling

**Reply by Mogball** [reply]
Not something being considered at the moment

**Reply by daydun** [reply]
@Mogball would you mind sharing an example that illustrates the issue? I don't quite understand why parsing would need to rely on typing when languages such as rust don't appear to have the same issue as far as I understand.

**Reply by Omicronlawful** [reply]
@Mogball why is it even using lazy parsing. Wouldn’t it make more sense to compile it into something it could easlily work with. Parsing while running can’t be efficient

**Reply by daydun** [reply]
Lazy parsing does not mean it's parsed at runtime. It means it's parsed in multiple stages during compile-time.

**Reply by Omicronlawful** [reply]
> Lazy parsing does not mean it's parsed at runtime. It means it's parsed in multiple stages during compile-time.

A good question would be why the parser needs indents in the first place to know what’s a body. Everything after a { is a body. I thought the indents are used to determine the end of the body. But this would be indicated by }.

**Reply by daydun** [reply]
Everything after a `{` is not body. The body eventually ends with a `}`. But you can't just find the next `}` and call that the end, since the curly bracket could be inside of a string for example, or it could be part of a register_passable constructor.

Though I still don't understand what would be so complicated about determining where the body ends with brackets. Maybe I'm missing something.

**Reply by Omicronlawful** [reply]
It’s oversimplified, but counting brackets and checking for escapes isn’t that much harder than counting indents(checks needed for tab indent and space indent and the amount of spaces). I don’t get the benefit of skipping parts to parse them later. I would parse every file and process the parsed input as I need to. It needs to be parsed anyway so why reading the same file multiple times.

**Comment by ksandvik** [comment]
Just my opinion but with modern code editors this whole issue of not seeing indentations with spaces is mostly gone, guard lines, coloring of blocks et rest. Adding curly brackets would also make it less Pythonic and also suddenly make use of curly brackets for some other purpose more complex concerning tokenization.

**Reply by Omicronlawful** [reply]
It definitely makes it easier, but just a single space can cause problems and it can be avoided. For example when you delete a block of code you indentations are messed up. In other languages you just reformat the code and it’s fine, but that does not work in python because it’s essential. And of course I’m not suggestion that to be a default behavior, but a selectable option. Just like you could write python like code, but also use struct and fn which looks more like rust code than python.

**Comment by ksandvik** [comment]
I see the point. I use black auto-save when formatting so it mostly handle such cases. I do think modern editors have made it much easier to avoid block notation in computer languages. Also, if this ever will happen, it should be an option as the idea of Mojo is to be a super-set of Python, you for example change the file postfix or use a simple translation tool to get going. Having the syntax very different will make Python coders very uncomfortable.

**Reply by Omicronlawful** [reply]
My first to words of this discussion where "The option". I doubt that this will make people uncomfortable, because most of the programming languages use {}, but I suggest this as a optional feature ;D

**Comment by lattner** [comment]
Hi all,

This suggestion cuts directly against or goals for Mojo, which is to be a member of the Python family.  Thank you for your suggestions, but our goal isn't to design a new language from first principles (been there done that :-) ), it is to lift an existing ecosystem.  We are also not adding general syntactic sugar, we are focused on core systems programming features that Python lacks.

-Chris
